File: conversions/is-primitive/elide-is-primitive.rb
#!/usr/bin/env ruby

# Ruby Integration Example for elide-is-primitive
#
# This demonstrates calling the TypeScript primitive type checker from Ruby
# using Elide's polyglot capabilities.
#
# Benefits:
# - One type checker shared across TypeScript and Ruby
# - Consistent primitive validation across microservices
# - Zero dependencies in both languages
# - Perfect for Rails models, validation, serialization

puts "=== Ruby Calling TypeScript is-primitive ===\n"

# NOTE: Exact syntax depends on Elide's Ruby polyglot API
# Assuming: is_primitive = Elide.require('./elide-is-primitive.ts')

# Example 1: Rails Model Validation
# class User < ApplicationRecord
#   validate :primitive_attributes
#
#   def primitive_attributes
#     [:name, :email, :age].each do |attr|
#       value = send(attr)
#       unless is_primitive.default(value)
#         errors.add(attr, "must be a primitive type")
#       end
#     end
#   end
# end

# Example 2: JSON Serialization Helper
# module SerializationHelper
#   def self.can_serialize?(value)
#     is_primitive.default(value)
#   end
#
#   def self.prepare_for_json(hash)
#     hash.select { |k, v| can_serialize?(v) }
#   end
# end
#
# data = { name: "John", age: 30, callback: proc {} }
# serializable = SerializationHelper.prepare_for_json(data)
# puts "Serializable: #{serializable}"
# puts ""

# Example 3: Cache Key Generation
# class CacheKeyGenerator
#   def self.generate(params)
#     primitive_params = params.select { |k, v|
#       is_primitive.default(v)
#     }
#     primitive_params.sort.to_s
#   end
# end
#
# params = { id: 1, name: "test", obj: {} }
# key = CacheKeyGenerator.generate(params)
# puts "Cache key: #{key}"
# puts ""

# Example 4: Deep Clone Optimization
# class DeepClone
#   def self.clone(obj)
#     if is_primitive.default(obj)
#       obj  # Primitives don't need cloning
#     elsif obj.is_a?(Hash)
#       obj.transform_values { |v| clone(v) }
#     elsif obj.is_a?(Array)
#       obj.map { |item| clone(item) }
#     else
#       obj.dup rescue obj
#     end
#   end
# end
#
# data = { name: "test", count: 5, nested: { val: 10 } }
# cloned = DeepClone.clone(data)
# puts "Cloned: #{cloned}"
# puts ""

# Example 5: API Parameter Validator
# class APIValidator
#   def self.validate_primitives(params, required_fields)
#     errors = []
#     required_fields.each do |field|
#       if params.key?(field)
#         unless is_primitive.default(params[field])
#           errors << "#{field} must be primitive"
#         end
#       end
#     end
#     errors
#   end
# end
#
# params = { name: "John", age: 30, settings: {} }
# errors = APIValidator.validate_primitives(params, [:name, :age, :settings])
# puts "Errors: #{errors}"
# puts ""

puts "Real-world use cases:"
puts "- Rails model validation"
puts "- JSON serialization helpers"
puts "- Cache key generation"
puts "- Deep clone optimization"
puts "- API parameter validation"
puts "- Type guards for primitives"
puts ""

puts "Problem Solved:"
puts "Before: Different primitive checks in TypeScript vs Ruby"
puts "After: One Elide implementation = consistent behavior"
puts ""

puts "When Elide Ruby API is ready:"
puts "  is_primitive = Elide.require('./elide-is-primitive.ts')"
puts "  result = is_primitive.default(5)  # true"

File: converted/utilities/is-primitive/elide-is-primitive.py
#!/usr/bin/env python3
"""
Python Integration Example for elide-is-primitive

This demonstrates calling the TypeScript primitive type checker from Python
using Elide's polyglot capabilities.

Benefits:
- One type checker shared across TypeScript and Python
- Consistent primitive validation across services
- Zero dependencies in both languages
- Perfect for data validation, serialization, API validation
"""

print("=== Python Calling TypeScript is-primitive ===\n")

# NOTE: Exact syntax depends on Elide's Python polyglot API
# Assuming: from elide import require
#           is_primitive = require('./elide-is-primitive.ts')

# Example 1: Basic Type Validation
# def validate_types():
#     """Check various types"""
#     test_values = [
#         5, "hello", True, None, [], {}, 
#         lambda x: x, 3.14, False, ""
#     ]
#     
#     print("Type validation:")
#     for val in test_values:
#         result = is_primitive.default(val)
#         print(f"  {repr(val)}: {'primitive' if result else 'object/function'}")
#     print()

# Example 2: API Input Validation
# class APIValidator:
#     @staticmethod
#     def validate_primitive_fields(data, fields):
#         """Validate that specified fields contain primitives"""
#         errors = []
#         for field in fields:
#             if field in data:
#                 if not is_primitive.default(data[field]):
#                     errors.append(f"{field} must be primitive type")
#         return errors
#
# data = {"name": "John", "age": 30, "settings": {}}
# errors = APIValidator.validate_primitive_fields(data, ["name", "age", "settings"])
# print(f"Validation errors: {errors}")
# print()

# Example 3: Serialization Check
# def can_serialize_directly(value):
#     """Check if value can be serialized without transformation"""
#     return is_primitive.default(value)
#
# print("Serialization check:")
# values = [42, "text", None, [], {}]
# for val in values:
#     can_ser = can_serialize_directly(val)
#     print(f"  {repr(val)}: {'direct' if can_ser else 'needs transform'}")
# print()

# Example 4: Deep Clone Detection
# class CloneHelper:
#     @staticmethod
#     def clone(obj):
#         """Clone object or return primitive directly"""
#         if is_primitive.default(obj):
#             return obj  # Primitives don't need cloning
#         elif isinstance(obj, dict):
#             return {k: CloneHelper.clone(v) for k, v in obj.items()}
#         elif isinstance(obj, list):
#             return [CloneHelper.clone(item) for item in obj]
#         else:
#             return obj
#
# data = {"name": "test", "count": 5, "nested": {"val": 10}}
# cloned = CloneHelper.clone(data)
# print(f"Cloned: {cloned}")
# print()

# Example 5: Cache Key Generator
# def generate_cache_key(params):
#     """Generate cache key from primitive params only"""
#     primitive_params = {
#         k: v for k, v in params.items() 
#         if is_primitive.default(v)
#     }
#     return str(sorted(primitive_params.items()))
#
# params = {"id": 1, "name": "test", "callback": lambda: None}
# cache_key = generate_cache_key(params)
# print(f"Cache key: {cache_key}")
# print()

print("Real-world use cases:")
print("- API input validation")
print("- Serialization checks for JSON/DB")
print("- Deep clone optimization")
print("- Cache key generation")
print("- Type guards for primitives")
print("- Data sanitization")
print()

print("Problem Solved:")
print("Before: Different primitive checks in TypeScript vs Python")
print("After: One Elide implementation = consistent behavior")
print()

print("When Elide Python API is ready:")
print("  from elide import require")
print("  is_primitive = require('./elide-is-primitive.ts')")
print("  result = is_primitive.default(5)  # True")
